proposals/0529-th-quotes-as-patterns.rst

```diff
@@ -0,0 +1,356 @@
+
+  =>
+
+  f (LamE [VarP __x0] (VarE __x1) | nameBase __x0 == "x" && __x0 == __x1 = ...
+
+This gets especially interesting with multiple scopes::
+
+  f [| (\x -> x, \x -> x) |] = ...
+
+  =>
+
+  f (TupE [ Just (LamE [VarP __x0] (VarE __x1))
+          , Just (LamE [VarP __x2] (VarE __x3))
+          ])
+    | nameBase __x0 == "x" && __x0 == __x1
+    | nameBase __x0 == "x" && __x2 == __x3
+    = ...
+
+Note how ``__x0`` is related to ``__x1`` and  ``__x2`` likewise to ``__x3``, but the former two are *not* related to the latter two.
+This respects the two independent scopes.
+
+This is perhaps convenient, but it rather baroque.
+It is also unclear whether the ``nameBase _ == "x"`` is useful in practice, or whether it is better to just "bake in" alpha equivalence and not care whether the local variable is an "x" or not provided the usage lines up with the binding.
+
+Finally, the non-linear patterns trick is not a true dual because it merely checks whether the variables "ended up" being the same *once the action is run*, rather than pattern matching on the action *itself*.
+De Bruijn indices encode actions in a way that makes equality of the easily decidable, for example, and thus would be a less hacky solution.
+
+---------
+
+Both alternatives are tempting, but I rather wait for more research on patterns, optics, and "dualizing" ``Applicative`` and effects like ``Quote`` in general, so we can better understand the theory of what's going on.
+If and when we understand the lay-of-the-land better, we can make some new extensions and deprecate the old ones accordingly.
+
+Unresolved Questions
+--------------------
+
+None at this time.
+
+Related Work
+------------
+
+Examples of langauges with also implement this feature
+
+Scheme
+~~~~~~
+
+- ``syntax-rules``, see https://docs.racket-lang.org/guide/pattern-macros.html#%28part._define-syntax_and_syntax-rules%29 .
+- ``syntax-case``, see https://docs.racket-lang.org/guide/syntax-case.html .
+
+Racket
+~~~~~~
+
+``syntax-parse`` is a more powerful alternative to ``syntax-case``.
+See:
+
+- Docs: https://docs.racket-lang.org/syntax/Parsing_Syntax.html
+
+- Paper: `"Fortifying Macros" by Culpepper and Felleisen <https://www2.ccs.neu.edu/racket/pubs/icfp10-cf.pdf>`_
+
+There's a Haskell implementation of ``syntax-parse``'s core as part of the Crucible language's `concrete syntax <https://github.com/GaloisInc/crucible/blob/master/crucible-syntax/src/Lang/Crucible/Syntax/ExprParse.hs>`_.
+
+Rust
+~~~~
+
+- Pattern-based macros (like ``syntax-rules``) https://doc.rust-lang.org/reference/macros-by-example.html .
+
+- Procedural macros don't have great pattern matching, but https://github.com/nrc/proc-macro-rules is an library-level experiment to try to bridge the gap.
+
+
+Idris
+~~~~~
+
+- https://davidchristiansen.dk/pubs/type-directed-elaboration-of-quasiquotations.pdf
+
+Lean
+~~~~
+
+- https://dl.acm.org/doi/pdf/10.1145/3110278
+
+- https://arxiv.org/pdf/2001.10490.pdf
+
+Future Work
```

Comment:
Would it make sense to have something like this for Typed Template Haskell? How far off is such a thing?

Reply:
Don't think it is too bad, but not sure. Added as future work.